src/Adapter/Language/CommandHandler/AbstractLanguageHandler.php

```diff
@@ -0,0 +1,205 @@
+                throw new CopyingNoPictureException(
+                    sprintf('An error occurred while copying "No Picture" image to product directory'),
+                    CopyingNoPictureException::PRODUCT_IMAGE_COPY_ERROR
+                );
+            }
+
+            if (!ImageManager::resize($temporaryImage, _PS_IMG_DIR_ . 'c/' . $imageName, $imageWidth, $imageHeight)) {
+                throw new CopyingNoPictureException(
+                    sprintf('An error occurred while copying "No Picture" image to category directory'),
+                    CopyingNoPictureException::CATEGORY_IMAGE_COPY_ERROR
+                );
+            }
+
+            if (!ImageManager::resize($temporaryImage, _PS_IMG_DIR_ . 'm/' . $imageName, $imageWidth, $imageHeight)) {
+                throw new CopyingNoPictureException(
+                    sprintf('An error occurred while copying "No Picture" image to brand directory'),
+                    CopyingNoPictureException::BRAND_IMAGE_COPY_ERROR
+                );
+            }
+        }
+
+        unlink($temporaryImage);
+    }
+
+    /**
+     * @param int $languageId
+     * @param string $newImagePath
+     * @param string $imageDir
+     */
+    protected function uploadImage($languageId, $newImagePath, $imageDir)
+    {
+        $temporaryImage = tempnam(_PS_TMP_IMG_DIR_, 'PS');
+        if (!$temporaryImage) {
+            return;
+        }
+
+        if (!move_uploaded_file($newImagePath, $temporaryImage)) {
+            return;
+        }
+
+        // Evaluate the memory required to resize the image: if it's too much, you can't resize it.
+        if (!ImageManager::checkImageMemoryLimit($temporaryImage)) {
+            throw new LanguageImageUploadingException(
+                'Due to memory limit restrictions, this image cannot be loaded. Increase your memory_limit value.',
+                LanguageImageUploadingException::MEMORY_LIMIT_RESTRICTION
+            );
+        }
+
+        // Copy new image
+        if (!ImageManager::resize($temporaryImage, _PS_IMG_DIR_ . $imageDir . $languageId . '.jpg')) {
+            throw new LanguageImageUploadingException(
+                'An error occurred while uploading the image. Check your directory permissions.',
+                LanguageImageUploadingException::UNEXPECTED_ERROR
+            );
+        }
+
+        if (file_exists(_PS_LANG_IMG_DIR_ . $languageId . '.jpg')) {
+            $shopId = Context::getContext()->shop->id;
+            $currentFile = _PS_TMP_IMG_DIR_ . 'lang_mini_' . $languageId . '_' . $shopId . '.jpg';
+
+            if (file_exists($currentFile)) {
+                unlink($currentFile);
+            }
+        }
+
+        unlink($temporaryImage);
+    }
+
+    /**
+     * @param int $languageId
+     * @param int[] $shopAssociation
+     */
+    protected function associateWithShops($languageId, array $shopAssociation)
```

Comment:
@matks this is pretty reusable method, it will definitely be used for other handlers as well? do you think we can extract it to let's say `trait` or some service?

Reply:
OK for trait 👍 
I know @eternoendless does not like them, but I think it's because he has seen lot of traits badly used.
Let's change his mind 😄

Reply:
Fun thing: Tomas PR has exactly the same idea: https://github.com/PrestaShop/PrestaShop/pull/11894/files#diff-306818ccc29ec93be5d17

Reply:
yeah, since there are a lot of common methods for object models, we thought that  `AbstractObjectModelLegacyHandler` would be a good place for that, wdyt? :)